# Project-AI Algorithms
### Noémie Mazepa, Aymeric Martin & Lucas Mesa Casellas

## **Problem Description**

The problem is that a biscuit factory wants to produce several types of Christmas biscuits on the same roll of dough. Each type of biscuits has a different size and a different price and the objective of this problem is to maximize the production of those biscuits and therefore ensure the best profit for the factory by finding an optimized solution for the biscuit placement on the dough roll.  

Let's describe the different elements of this problem in detail:

#### **Dough roll charateristics**:

1. **Length**: the dough roll has a fixed length of 500 units.

2. **Defects**: the dough roll contains a lot of defects along its length.

3. **Biscuits**: There is a list of biscuits positioned all along the roll.

4. **Penalties**: if a portion of the dough roll is unused, the final value gets a penalty of -1 for each unused unit to reflect wasted material and profit loss.


#### **Defects charateristics**:

1. **Position**: each defect has a specific position on the roll.

2. **Classes**: defects are categorized into three classes: a, b, and c.

**->** The number and class of the defects directly affect the placement of biscuits because each biscuit type has certain thresholds for the number of defects allowed on the biscuit.

Here is a representation of the defects placed all along the dough roll:

![Defects Representation](DefectsPositions.png "Defects Representation")



#### **Biscuits charateristics**:

The factory produces four types of biscuits, each with the following characteristics:
- **Biscuit 0**:

        Length: 4
        Value: 3
        Defect Thresholds: {a: 4, b: 2, c: 3}

- **Biscuit 1**:

        Length: 8
        Value: 12
        Defect Thresholds: {a: 5, b: 4, c: 4}

- **Biscuit 2**:

        Length: 2
        Value: 1
        Defect Thresholds: {a: 1, b: 2, c: 1}

- **Biscuit 3**:

        Length: 5
        Value: 8
        Defect Thresholds: {a: 2, b: 3, c: 2}



#### **Constraints**:

1. **Placement**:

- Biscuits need to be placed at integer positions on the roll.
- No biscuits can overlap.

2. **Defect Tolerance**:

- The number of defects in the positions covered by a biscuit should not be over its specific threshold for any defect class.

3. **Total Length**:

- The sum of the lengths of all placed biscuits should not be superiror to the total length of the roll.

4. **Penalties**:

- All the dough must be used, if there is a unit that is unused we give a penalty of -1 to the total value.



#### **Objective Function**:

To ensure the best profit for the factory, we need to maximize the following function:

- **Total Value=Sum of Biscuit Values − Unused Length**

This can be calculated be summing the price of each biscuit type multiplied by its number present on the dough minus all of the penalties given.

## **Implementation of 2 Search Algorithms**

In order to find an optimal solution for the biscuit placement on the dough roll while maximizing the profit for the factory, we will implement **2 different search algorithms** to solve the problem. We will then evaluate their performance and compare their results.

For this project, we have chosen to implement the following algorithms:

#### **A Star**:

A* is an informed search algorithm that strategically balances path cost and future potential by evaluating nodes using a cost function:

**f(x)=g(x)+h(x)**

Where:
- **g(x)**: The actual cost to reach the current state.
- **h(x)**: A heuristic estimating the cost to reach the goal from the current state.

A* explores the most promising paths first, using heuristics that prioritize solutions with the highest potential profit while minimizing penalties.

**Implementation Details**:

- **Node Representation**: Each node includes the current position on the dough roll, the list of biscuits already placed, and the total value accumulated.

- **Heuristic Design**: We use the maximum value-per-unit between all the biscuit types to estimate the remaining potential profit (h(x)). This way we ensure the algorithm focuses on maximizing the value while minimizing penalties.

- **Node Expansion**: Valid biscuits are determined based on constraints like as defect thresholds and available dough length. Penalty states are added for wasted dough units, which gives a cost of -1.

- **Priority Queue**: A* uses a priority queue (min-heap) to explore states with the lowest f(x) value first, ensuring an optimal solution is found.



#### **Constraint Satisfaction Problem (CSP)**:

CSP is a problem-solving algorithm where the objective is to find an assignment of variables that satisfies a set of constraints.
In this context:
- **Domains**: The list of biscuit types that can be placed at each position, considering the defect thresholds and remaining roll length.
- **Variables**: Positions of the biscuits placed on the dough roll.
- **Constraints**: Defect thresholds, no overlapping biscuits, and total roll length.

We will use **Backtracking Search**: we assign values to variables until one of them creates a conflicts (a constraint is violated) and then algorithm bakctracks: it goes back to the previous assignement and tries a different one.
We repeat the process until all variables are assigned with valid values: we reached the end of the dough roll.

We will use **Heuristics** as well:
- **MRV (Minimum Remaining Values)**: to select an unassigned variable, we run the MRV to choose the variable with the smallest domain.
- **Degree**: if there are ties during MRV (several variables have the smallest domain), we use the Degree heuristic to prioritize the variable that is involved in the most constraints.
- **LCV (Least Constraining Value)**:we use it to order the domain of a variable so that we test first the value that leaves the most options for the neighboring variables.   



#### **Why These Algorithms?**

**A Star**:
A* works well for this problem because it focuses on exploring the best solutions first. It balances meeting the constraints, like defects and roll length, with the goal of maximizing profit. The use of heuristics helps it manage large search spaces effectively.

**CSP**:
CSP fits this problem because it deals directly with constraints, which are a big part of the challenge here. Using techniques like constraint propagation and heuristics, it efficiently narrows down the possibilities while ensuring all placement rules are followed.



Out of all the possible approaches, A* and CSP stood out as the most interesting to study because they each bring a unique perspective to problem-solving.

A* is appealing because of its balance between cost-efficiency and future potential, making it a great example of how heuristics can guide decision-making in complex optimization problems. It’s a practical choice for tackling real-world challenges where finding a good solution quickly is crucial.

CSP, on the other hand, is fascinating due to its structured approach to constraints. Since this problem is heavily constraint-based, studying CSP allows us to dive into how rules and restrictions can be managed effectively. Techniques like constraint propagation and heuristics make CSP both methodical and efficient, which is why it seemed like a great fit for exploring this type of problem.

Together, these two approaches offer a solid mix of optimization and constraint management, making them both practical and insightful to learn from.
    

#### **Performance Evaluation**:

We will compare both algorithms based on several criterias:

1. **Total Value**: The total value of biscuits placed.

Obviously, because we need to optimize the production and the total value, we will evaluate both algorithms on how high the total value found is.



2. **Dough Waste**: The quantity of dough units wasted/unused.

We will check how many penalties the solution received. The less penalties, the better, because as much as we want to optimize the profit we also don't want to waste too much dough.



3. **Execution Time**: How efficiently they find a solution.

We will compare the time it takes for the algorithm to find a solution. In real-time situations, we need to find the best solution quickly.

## **A\* Algorithm**

In [16]:
import pandas as pd
import numpy as np
from heapq import heappush, heappop
import time
from collections import Counter

In [17]:
df_defects=pd.read_csv("defects.csv", sep=',')
df_defects.head()

x class
0  355.449335     c
1   92.496236     a
2  141.876795     c
3  431.833902     c
4  435.028461     c

In [18]:
class Biscuit:
    biscuit_types = {
        0: {'size': 4, 'price': 3, 'threshold_defects': {'a': 4, 'b': 2, 'c': 3}},
        1: {'size': 8, 'price': 12, 'threshold_defects': {'a': 5, 'b': 4, 'c': 4}},
        2: {'size': 2, 'price': 1, 'threshold_defects': {'a': 1, 'b': 2, 'c': 1}},
        3: {'size': 5, 'price': 8, 'threshold_defects': {'a': 2, 'b': 3, 'c': 2}}
    }

    def __init__(self, type, id=0):
        """
        Initializes a Biscuit with the specified type and optional id.

        Args:
            type (int): The type of biscuit. It must be a key in `biscuit_types`.
            id (int, optional): A unique identifier for the biscuit.

        Raises:
            ValueError: If the provided type is not in `biscuit_types`.

        Attributes:
            type (int): The type of the biscuit.
            id (int): The unique identifier for the biscuit.
            size (int): The size of the biscuit.
            price (int): The price of the biscuit.
            threshold_defects (dict): Thresholds for defect.
        """
        if type not in self.biscuit_types:
            raise ValueError(f"Biscuit type {type} is not valid.")

        self.type = type
        self.id = id

        #for the rest of the informations, the function will
        # automatically fetch them from the dictionnary of the class where all the informations are kept for each type.
        self.size = self.biscuit_types[type]['size']
        self.price = self.biscuit_types[type]['price']
        self.threshold_defects = self.biscuit_types[type]['threshold_defects']

    def __repr__(self):
        """
        Returns a string representation of the Biscuit object.

        Returns:
            str: A string to represent the biscuit.
        """
        return f"Biscuit(id={self.id}, type={self.type})"





In [19]:
class Node:
    def __init__(self, position, biscuits, value, g, h):
        """
        Initializes a Node object with the given attributes.

        Args:
            position (tuple): The position we are on the dough.
            biscuits (list): The list of biscuits associated placed on the dough so far.
            value (any): The accumulated price of the dough so far.
            g (float): The cost to reach this node from the start node.
            h (float): The estimated cost to reach the goal node from this node (heuristic).

        Attributes:
            position (tuple): The position we are on the dough.
            biscuits (list): The list of biscuits associated placed on the dough so far.
            value (any): The accumulated price of the dough so far.
            g (float): The cost to reach this node from the start node.
            h (float): The estimated cost to reach the goal node from this node (heuristic).
            f (float): Total cost (f = g + h), used to determine priority in algorithms like A*.
        """
        self.position = position
        self.biscuits = biscuits
        self.value = value
        self.g = g
        self.h = h
        self.f = g + h


    def __lt__(self,node):
        """
        This method compares the node to another based on their total path cost.

        Args:
            node (Node): the other we want to to compare the actual node to

        Returns:
            bool: returns true if the actual node's total path cost is lesser than the other node's total path cost, return false otherwise
        """
        return self.f < node.f #this method is useful for the priority queue as it knows how to compare its different elements


In [20]:
class DoughRoll:
    def __init__(self, df_defects,length=500):
        """
        Initializes the dough roll.

        Args:
            df_defects (DataFrame): A DataFrame containing defect data with columns 'x' (position) and 'class' (defect type).
            length (int, optional): The total length of the dough roll. Defaults to 500.

        Attributes:
            length (int): The length of the dough roll.
            defects (dict): A dictionary mapping positions to lists of defect types.
        """
        self.length=length
        self.defects=self.get_defects(df_defects)

    def get_defects(self, df_defects):
        """
        Processes the defect DataFrame into a dictionary for easier access.

        Args:
            df_defects (DataFrame): A DataFrame containing the defect data.

        Returns:
            dict: A dictionary mapping positions to lists of defect types.
        """
        defects_dict = df_defects.groupby('x')['class'].apply(list).to_dict()
        return defects_dict

    def can_place_biscuit(self, biscuit, position):
        """
        Checks if a biscuit can be placed at a specific position.

        Args:
            biscuit (Biscuit): The biscuit to place.
            position (int): The position where the biscuit is to be placed.

        Returns:
            bool: True if the biscuit can be placed, False otherwise.
        """
        return self.check_defects(biscuit, position) and self.check_position_int(position) and self.check_position_valid(biscuit, position)

    def check_defects(self, biscuit, position):
        """
        Verifies that defects at a position do not exceed the biscuit's thresholds.

        Args:
            biscuit (Biscuit): The biscuit to check.
            position (int): The position to check for defects.

        Returns:
            bool: True if the defects are within acceptable thresholds, False otherwise.
        """
        thresholds = biscuit.threshold_defects
        count_defect_a = 0
        count_defect_b = 0
        count_defect_c = 0

        start_pos = position
        end_pos = position + biscuit.size
        #we set an interval for the entire biscuit length starting at the given position and we get all the defects placed within this interval
        defects = {pos: classes for pos, classes in self.defects.items() if start_pos <= pos < end_pos}

        #for every defect in the interval, we look up its class and add it to the counter.
        #if a count is over the assigned threshold, the biscuit can't be placed.
        for defect_classes in defects.values():
            for defect_class in defect_classes:
                if defect_class == 'a':
                    count_defect_a += 1
                    if count_defect_a > thresholds.get('a', 0):
                        return False
                elif defect_class == 'b':
                    count_defect_b += 1
                    if count_defect_b > thresholds.get('b', 0):
                        return False
                elif defect_class == 'c':
                    count_defect_c += 1
                    if count_defect_c > thresholds.get('c', 0):
                        return False

        return True

    def check_position_int(self, position):
        """
        Checks if the position is an integer.

        Args:
            position (int): The position to check.

        Returns:
            bool: True if the position is an integer, False otherwise.
        """
        return type(position)==int

    def check_position_valid(self, biscuit, position):
        """
        Checks if a biscuit fits on the dough roll starting at the given position.

        Args:
            biscuit (Biscuit): The biscuit to place.
            position (int): The starting position.

        Returns:
            bool: True if the biscuit fits, False otherwise.
        """
        return biscuit.size + position<=self.length


    def get_biscuits(self, position):
        """
        Retrieves a list of biscuits that can be placed at the given position.

        Args:
            position (int): The starting position.

        Returns:
            list: A sorted list of biscuits by their value-to-size ratio in descending order.
        """
        remaining_length = self.length - position
        list_biscuits=[]
        for i in range(4): #for each type of biscuit we check if the biscuit fits the constraints; if yes we add it to the list.
            biscuit=Biscuit(i)
            if biscuit.size <= remaining_length and self.can_place_biscuit(biscuit, position):
                list_biscuits.append(biscuit)
        list_biscuits.sort(key=lambda b: b.price / b.size, reverse=True)#we order the list to place the best biscuits first.
        return list_biscuits

    def is_goal(self,position):
        """
        Checks if the goal (end of the dough roll) is reached.

        Args:
            position (int): The current position.

        Returns:
            bool: True if the position is at or beyond the roll length, False otherwise.
        """
        return position>=self.length

    def heuristics(self, position):
        """
        Estimates the potential maximum profit from the remaining dough roll.

        Args:
            position (int): The current position.

        Returns:
            float: The estimated maximum profit.
        """
        remaining_length = self.length - position
        max_profit_per_unit = max(b.price / b.size for b in [Biscuit(i) for i in range(4)])
        return remaining_length * max_profit_per_unit


    #A* search
    def a_star_search(self):
        """
        Performs an A* search to find the optimal arrangement of biscuits on the dough roll.

        Returns:
            Node: The best solution node containing the arrangement with maximum value.
        """
        self.start_time_a_star = time.time()
        self.penalties = 0
        self.biscuit_counter = Counter()

        initial_node=Node(position=0,biscuits=[],value=0,g=0,h=self.heuristics(0)) #initialization of the first node, we start at position 0.
        frontier = [] #the frontier is the list where all the nodes to visit will be placed.
        heappush(frontier, initial_node)
        best_solution = None #because the first solution returned isn't necessarily the best, we need to keep in memory the best solution encountered so far and compare everytime a branch reaches its end.
        visited_states = np.full(self.length, -float('inf'))#we keep in memory the values of the dough at certain positions to avoid exploring unpromising nodes

        while frontier: #we keep visiting nodes until there none to visit anymore.
            current_node = heappop(frontier)
            if self.is_goal(current_node.position): #if we reach the end of the dough (the dough is full) we try to see if it is better than our last best solution.

                if best_solution is None or current_node.value > best_solution.value:
                    best_solution = current_node
                continue

            if visited_states[current_node.position] >= current_node.value: #if the value at a certain position is esser than the maximum value at this position, the node is suboptimal and we don't even want to expand the branch anymore.
                continue
            visited_states[current_node.position] = current_node.value

            valid_biscuits=self.get_biscuits(current_node.position) #we get all the "childs": all the next biscuits we can place
            if valid_biscuits:
                for biscuit in valid_biscuits:
                    new_position = current_node.position + biscuit.size
                    new_biscuits = current_node.biscuits + [(biscuit, current_node.position)]
                    new_value = current_node.value + biscuit.price
                    new_node = Node(position=new_position,biscuits=new_biscuits,value=new_value,g=current_node.g + biscuit.size,h=self.heuristics(new_position))
                    heappush(frontier, new_node)
            else:
                if current_node.position+1<=self.length: #if there are no biscuits that can be placed, we give a penalty.
                    self.penalties += 1
                    penalty_value = current_node.value - 1
                    penalty=Node(position=current_node.position + 1,biscuits=current_node.biscuits,value=current_node.value - 1,g=current_node.g + 1,h=self.heuristics(current_node.position + 1))
                    heappush(frontier, penalty)

        self.end_time_a_star = time.time()
        self.solution_a_star = best_solution #we keep the best solution found
        return best_solution


    def calculate_number_penalty(self):
        """
        Calculates the positions that remain unused on the dough roll.

        Returns:
            list: A list of unused positions.
        """
        covered_positions = set()
        for biscuit, pos in self.solution_a_star.biscuits:
            for i in range(pos, pos+biscuit.size):
                covered_positions.add(i)
        penalty_positions=[pos for pos in range(self.length) if pos not in covered_positions]
        return penalty_positions

    def display_perfromance_a_star(self):
        """
        Displays the performance metrics of the A* search, including execution time, biscuit breakdown,
        penalties, and total value.
        """
        for biscuit, _ in self.solution_a_star.biscuits:
                self.biscuit_counter[biscuit.type] += 1
        penalty_positions=self.calculate_number_penalty()
        # Print performance metrics
        print(f"Execution Time: {self.end_time_a_star - self.start_time_a_star:.2f} seconds")
        print(f"Biscuit Breakdown: {dict(self.biscuit_counter)}")
        value_without_penalties=sum(count * Biscuit.biscuit_types[b_type]['price'] for b_type, count in self.biscuit_counter.items())
        print(f"Total Value without penalties: {value_without_penalties}")
        print(f"Number of Penalties: {len(penalty_positions)}")
        print(f"Penalty Positions: {penalty_positions}")
        print(f"Total Value: {self.solution_a_star.value}")



In [21]:
dough_roll = DoughRoll(df_defects)

solution = dough_roll.a_star_search()
dough_roll.display_perfromance_a_star()
print("\n")

if solution:
    print("Best solution found:")
    for biscuit, position in solution.biscuits:
        print(f"Biscuit {biscuit.type} placed at position {position}")
else:
    print("No valid solution found.")

Execution Time: 1.95 seconds
Biscuit Breakdown: {1: 33, 3: 36, 2: 7, 0: 9}
Total Value without penalties: 718
Number of Penalties: 6
Penalty Positions: [97, 113, 157, 166, 183, 184]
Total Value: 712


Best solution found:
Biscuit 1 placed at position 0
Biscuit 1 placed at position 8
Biscuit 1 placed at position 16
Biscuit 3 placed at position 24
Biscuit 3 placed at position 29
Biscuit 2 placed at position 34
Biscuit 0 placed at position 36
Biscuit 1 placed at position 40
Biscuit 2 placed at position 48
Biscuit 0 placed at position 50
Biscuit 1 placed at position 54
Biscuit 3 placed at position 62
Biscuit 1 placed at position 67
Biscuit 0 placed at position 75
Biscuit 3 placed at position 79
Biscuit 3 placed at position 84
Biscuit 1 placed at position 89
Biscuit 1 placed at position 98
Biscuit 2 placed at position 106
Biscuit 3 placed at position 108
Biscuit 0 placed at position 114
Biscuit 0 placed at position 118
Biscuit 1 placed at position 122
Biscuit 1 placed at position 130
Biscui

## **CSP Algorithm**

In [22]:
from collections import Counter
import pandas as pd
import random
import time


class DoughRollCSP:

    def get_defects(self, df_defects):
        # Group defects by their x-coordinate and collect their 'class' values into lists
        defects_dict = df_defects.groupby('x')['class'].apply(list).to_dict()
        return defects_dict

    def __init__(self, df_defects, length=500, variable_heuristic='default', value_heuristic='default'):
        self.length = length
        self.defects = self.get_defects(df_defects)
        self.variable_heuristic = variable_heuristic
        self.variables = list(range(length))  # Each position on the dough is a variable
        self.domains = {pos: list(range(4)) for pos in self.variables}  # Biscuit types (0-3) as domain

    def is_consistent(self, biscuit, position, assignment):
        """
        Check if placing the biscuit at the given position satisfies all constraints.
        """
        # Check if the biscuit stays within bounds
        if position + biscuit.size > self.length:
            return False

        # Check defect constraints
        if not self.check_defects(biscuit, position):
            return False

        # Check overlap with already-placed biscuits
        for placed_biscuit, placed_position in assignment:
            if self.check_overlap(biscuit, position, placed_biscuit, placed_position):
                return False

        return True  # All constraints satisfied


    def check_overlap(self, biscuit_i, position_i, biscuit_j, position_j):
        """
        Check if two biscuits overlap based on their positions and sizes.
        """
        # Biscuits overlap if their positions and sizes intersect
        return not (position_i + biscuit_i.size <= position_j or position_j + biscuit_j.size <= position_i)

    def check_defects(self, biscuit, position):
        """
        Check if placing the biscuit at the given position violates defect thresholds.
        """
        thresholds = biscuit.threshold_defects  # Thresholds for defects per biscuit type
        count_defect_a = 0  # Initialize counters for defect types
        count_defect_b = 0
        count_defect_c = 0

        # Determine the range of positions the biscuit will occupy
        start_pos = position
        end_pos = position + biscuit.size

        # Extract defects that fall within the biscuit's range
        defects = {}

        # Iterate through all defect positions and classes
        for pos, classes in self.defects.items():
            # Check if the position is within the biscuit's range
            if start_pos <= pos < end_pos:
                defects[pos] = classes

        # Count defects and check if they exceed the thresholds
        for defect_classes in defects.values():
            for defect_class in defect_classes:
                if defect_class == 'a':
                    count_defect_a += 1
                    if count_defect_a > thresholds.get('a', 0):  # Exceeds threshold for 'a'
                        return False
                elif defect_class == 'b':
                    count_defect_b += 1
                    if count_defect_b > thresholds.get('b', 0):  # Exceeds threshold for 'b'
                        return False
                elif defect_class == 'c':
                    count_defect_c += 1
                    if count_defect_c > thresholds.get('c', 0):  # Exceeds threshold for 'c'
                        return False

        return True  # No thresholds violated



    def backtrack(self, assignment, position):
        """
        Backtracking algorithm to solve the CSP with heuristics.
        Randomly selects a biscuit type for each position and backtracks if necessary.
        """
         # print(f"Current assignment: {assignment}, Position: {position}")  # Debug print
        if position >= self.length:  # If the end of the dough is reached, return the solution
            return assignment

        # one = [0, 1, 2, 3], two = [0, 1, 3, 2], three = [0, 2, 1, 3], four = [0, 2, 3, 1], five = [0, 3, 1, 2], six = [0, 3, 2, 1], seven = [1, 0, 2, 3] ,eight = [1, 0, 3, 2], nine = [1, 2, 0, 3], ten = [1, 2, 3, 0], eleven = [1, 3, 0, 2], twelve = [1, 3, 2, 0], thirteen = [2, 0, 1, 3], fourteen = [2, 0, 3, 1], fifteen = [2, 1, 0, 3], sixteen = [2, 1, 3, 0], seventeen = [2, 3, 0, 1], eighteen = [2, 3, 1, 0], nineteen = [3, 0, 1, 2], twenty = [3, 0, 2, 1], twenty_one = [3, 1, 0, 2], twenty_two = [3, 1, 2, 0], twenty_three = [3, 2, 0, 1], twenty_four = [3, 2, 1, 0]
        # Shuffle the biscuit types for randomness
        #random.shuffle(biscuit_types)  # Randomize the order of biscuits to try ////////////////////////////////
        order = [3, 1, 0, 2]
        # Try placing a random biscuit type at the current position
        for biscuit_types in order:
            biscuit = Biscuit(biscuit_types)
             # print(f"Trying Biscuit {biscuit_types} at position {position}, Domain size: {len(self.domains[position])}")
            if self.is_consistent(biscuit, position, assignment):
                # Try placing the biscuit
                assignment.append((biscuit, position))
                next_position = position + biscuit.size
                result = self.backtrack(assignment, next_position)
                if result is not None:
                    return result
                # Backtrack: remove the biscuit
                assignment.pop()

        # If no biscuit can be placed, move one unit forward and apply a penalty
        if position < self.length:
            return self.backtrack(assignment, position + 1)

        return None

    def calculate_unused_dough(self, assignment):
        """
        Calculate the total unused dough after placing biscuits.
        """
        used_positions = set()

        # Mark all positions occupied by biscuits in the assignment
        for biscuit, position in assignment:
            for i in range(position, position + biscuit.size):
                used_positions.add(i)

        # Calculate unused positions
        total_positions = set(range(self.length))
        unused_positions = total_positions - used_positions

        return len(unused_positions)  # Total number of unused positions

    def solve(self):
        """
        Solve the CSP using backtracking.
        """
        assignment = []  # Start with an empty assignment

        start_time = time.time()

        # Start backtracking with position 0
        solution = self.backtrack(assignment, 0)

        # End measuring time
        end_time = time.time()

        # Calculate total execution time
        execution_time = end_time - start_time


        if solution:
            total_value = sum(biscuit.price for biscuit, _ in solution)
            unused_dough = self.calculate_unused_dough(solution)
            print("Solution found:")
            for biscuit, position in solution:
                print(f"Biscuit {biscuit.type} placed at position {position}")
            print("Total value:", total_value)
            print(f"Unused dough: {unused_dough} ")
        else:
            print("No valid solution found.")


        print(f"Execution time: {execution_time:.4f} seconds")

    ################################# Heuristics part

    def select_unassigned_variable(self, assignment):
        """
        Select the next variable to assign based on heuristics.
        """
        unassigned = [var for var in self.variables if var not in [pos for _, pos in assignment]]
        if not unassigned:
            return None  # All variables are assigned

        print(f"Unassigned variables: {unassigned}")

        if self.variable_heuristic == 'MRV':
            # Minimum Remaining Values
            selected_var = min(unassigned, key=lambda var: len(self.domains[var]))
            print(f"MRV selected variable: {selected_var} with domain size {len(self.domains[selected_var])}")
            return selected_var

        elif self.variable_heuristic == 'DEG':
            # Degree Heuristic
            selected_var = max(unassigned, key=lambda var: self.get_degree(var, assignment))
            print(f"DEG selected variable: {selected_var} with degree {self.get_degree(selected_var, assignment)}")
            return selected_var

        elif self.variable_heuristic == 'MRV_DEG':
            # MRV with Degree Heuristic as a tie-breaker
            mrv_vars = [var for var in unassigned if len(self.domains[var]) == min(len(self.domains[v]) for v in unassigned)]
            selected_var = max(mrv_vars, key=lambda var: self.get_degree(var, assignment))
            print(f"MRV_DEG selected variable: {selected_var} with degree {self.get_degree(selected_var, assignment)}")
            return selected_var

        else:
            # Default: Pick the first unassigned variable
            selected_var = unassigned[0]
            print(f"Default selected variable: {selected_var}")
            return selected_var


    def get_degree(self, var, assignment):
        """
        Get the number of constraints affected by the variable.
        """
        return len([neighbor for neighbor in self.variables if neighbor != var and neighbor not in [pos for _, pos in assignment]])

     ############################

#CSP call
dough_roll_csp = DoughRollCSP(df_defects, length=500, variable_heuristic='default')
dough_roll_csp.solve()


Solution found:
Biscuit 1 placed at position 0
Biscuit 1 placed at position 8
Biscuit 3 placed at position 16
Biscuit 1 placed at position 21
Biscuit 3 placed at position 29
Biscuit 2 placed at position 34
Biscuit 0 placed at position 36
Biscuit 1 placed at position 40
Biscuit 2 placed at position 48
Biscuit 0 placed at position 50
Biscuit 3 placed at position 54
Biscuit 3 placed at position 59
Biscuit 1 placed at position 64
Biscuit 1 placed at position 72
Biscuit 1 placed at position 80
Biscuit 3 placed at position 88
Biscuit 0 placed at position 93
Biscuit 1 placed at position 98
Biscuit 0 placed at position 106
Biscuit 0 placed at position 111
Biscuit 0 placed at position 118
Biscuit 1 placed at position 122
Biscuit 3 placed at position 130
Biscuit 0 placed at position 135
Biscuit 0 placed at position 139
Biscuit 0 placed at position 143
Biscuit 3 placed at position 147
Biscuit 3 placed at position 152
Biscuit 3 placed at position 158
Biscuit 2 placed at position 164
Biscuit 3 plac

## **Comparison of the 2 algorithms**

### **Algorithm Results**

| Metric                             | A* Algorithm            | CSP Algorithm           |
|------------------------------------|-------------------------|-------------------------|
| **Execution Time**                 | 1.95 seconds            | 0.0256 seconds          |
| **Total Value (Before Penalties)** | 718                     | 705                     |
| **Total Value (After Penalties)**  | 712                     | 690                     |
| **Number of Penalties**            | 6                       | 15                      |






The A* and CSP algorithms were applied to the same problem but returned different results on the key performance metrics. The detailed comparison of the two approaches is presented below based on execution time, solution quality, and penalties.

- **Execution Time**: 
The CSP algorithm was considerably faster than A*, since it finished execution in just 0.0256 seconds compared to the 1.95 seconds that A* took. This is because the CSP algorithm simply tries to find any solution which satisfies the constraints without having to go through all the other solutions like A* does, searching for the most feasible solution.

- **Total Value**: 
Before the penalties were applied, the total value from the CSP algorithm solution stood at 705, against 718 for A*. This again shows that the simpler strategy for CSP can easily compete with the solution value.

However, after considering the penalties, A* had a higher total value, 712, compared to CSP, 702. This result indicates that A* is strong in including penalties and constraints in its search process for a more optimized final solution.

- **Number of Penalties**: 
A key point of divergence would be in penalty handling: CSP's algorithm incurred a total of 15 penalties, while that of A* had 6. That simply explains how CSP cannot promise an optimum because it only tries to satisfy constraints but does not aim at any more optimal. Meanwhile, the A* algorithm evaluates nodes one by one, causing fewer constraints.

- **Overall Analysis**: 
**A\* Algorithm**: Much slower, A* is great at finding an optimized solution, balancing raw value and penalties. It's more suited for problems that require a carefully refined final outcome where execution time is less critical.
**CSP Algorithm**: CSP is faster and effective for quickly solving problems with strict constraints but sacrifices optimality, incurring higher penalties. The utility of the algorithm will then be in giving a feasible solution fast if time is a priority.
In the end, the final decision comes to either A* or CSP based on what exactly one prioritizes about the problem at hand, either speed or solution quality.






## **Conclusion and Reflections**


#### **Challenges**:

- **A\*:**
1. The main challenge we encountered was that the solution returned first by the A* algorithm isn't necessarily the best solution. For a long time we were stuck with the final value 661 and no matter what we changed in the algorithm, this value stayed the same. It took us quite some time to realize, when talking to some other groups, that the best solution isn't the first one and that we should visit every single branch entirely and keep in memory to return the best solution. The reason why, the first solution isn't optimal like it is supposed to is the end goal. The problem with this problem is that the end goal isn't a specific configuration: it's just that we reached the end of the dough. This way, the "optimal solution" is only the one where we reach the end of the dough the fastest.
**->** What we did was that we expand all the nodes possible and look at all the different solutions possible and only keep the best one.

2. After we realized that, we tried to compile the search until the frontier was empty and all the solutions were found to get the best. However, there was a second problem that made the expansion and visit of the nodes would never finish or at least take over 10 minutes before it was interrupted. This is because to avoid visiting the same nodes several times, we used a set to keep in memory each of the visited nodes to ensure we don't visit them twice. We took inspiraion from the 8-square puzzle from Homework 1 where some configurations repeated themselves sometimes in the tree. However, it took us a long time to realize that no combination could repeat itself in different branches with this specific problem so this set was definitely unuseful.

**->** Instead we chose to keep in memory the best price for a given position and each time we visit a node, if for the same position, the value of the current node is lesser, we don't expand it because it could never be the best solution going forward. This makes the search go much faster and allows us to find the best solution.

- **CSP :**

1. Initial Challenge: The main problem with the CSP approach was that it searches for the first valid solution, not the best. Unlike A*, CSP does not naturally evaluate and compare all possible solutions to find the optimal one. Instead, it stops as soon as it finds a configuration that satisfies all constraints. This can result in a suboptimal total value.

2. Optimization Challenge: Since this was a limitation, we had to see whether the problem setup regarding CSP can be modified to return better solutions. The standard CSP framework is very good at quickly returning valid solutions. However, in order for it to optimize the sum of values, there would need to be major restructuring of the approach, which may involve adding additional constraints or running multiple runs with different setups. This will actually defeat the core mechanism of the CSP in satisfying constraints.

3. Memory Management: Unlike A*, CSP does not inherently remember visited nodes or require a significant amount of memory to handle the frontier or heuristic evaluations. However, making sure the constraints were efficient at every step required an overall careful tracking of biscuit positions and defect thresholds to avoid unnecessary recomputation.

4. Key Learning : CSP was much quicker in execution simply because it is lean and its goal is to reach a solution, not to explore all configurations or produce the "best" solution, according to some value. Hence, it is intrinsically fit for problems where it is acceptable to have any valid solution given, rather than optimization problems.


#### **Reflections**:


- **A\*:**
1. Advantages: A* is great for figuring out the most profitable way to produce biscuits by balancing production value and penalties. It’s perfect for identifying solutions that maximize profits while minimizing waste or defects. In biscuit production, this means A* can help find the best combination of steps to produce high-quality biscuits with fewer issues.

2. Disadvantages: The downside of A* is that it can be slow. Since it looks at all possible solutions to find the best one, it takes more time to run, especially with complex setups. For biscuit production, this might not work well when quick decisions are needed, like adapting to sudden changes in demand or production issues. Plus, A* uses a lot of memory, so managing that carefully is key to keeping things running smoothly.

3. Conclusion: A* is a great choice when the focus is on getting the best results, like maximizing profits and minimizing penalties, even if it takes a bit longer. It’s especially useful for planning and optimizing processes where quality and efficiency are more important than speed. To make A* work well, it’s important to think about how you’re exploring solutions and to keep an eye on memory usage so it doesn’t slow down too much.


- **CSP :**
1. Efficiency vs. Optimality: CSP is great at quickly finding a solution that meets the constraints, often with very little computation time. However, this speed comes with a tradeoff—it doesn’t always find the best solution. For example, if the goal is to maximize or minimize a value (like the biscuit value here), CSP might settle for a valid solution that isn’t the most optimal. This highlights its strength in solving problems fast but also shows its limitations when optimization is key.

2. Scalability: CSP’s simple and focused approach makes it highly scalable for problems with clear rules and limited possibilities. It works especially well when the constraints are strict, and any valid solution will do. However, as the constraints or the need for optimization get more complex, CSP might need significant adjustments, which could make it harder to use effectively.

3. Applicability to real-world problems: CSP fits perfectly for real-world tasks like scheduling, resource allocation, or layout planning—situations where the main goal is just to meet specific constraints. But it’s less suited for problems where you need to explore or refine multiple solutions to find the best one. CSP is a solid first step for quickly finding a feasible solution, but for fine-tuning or optimization, it’s often best paired with other techniques like A*.
